blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1033 @@
+            .put(msg, Blake2b256)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        Ok(cid)
+    }
+
+    fn state_get_actor(&self, addr: &Address, ts: &Tipset) -> Result<ActorState, Error> {
+        let state = StateTree::new_from_root(self.cs.db.as_ref(), ts.parent_state())
+            .map_err(Error::Other)?;
+        let actor = state.get_actor(addr).map_err(Error::Other)?;
+        actor.ok_or_else(|| Error::Other("No actor state".to_owned()))
+    }
+
+    fn messages_for_block(
+        &self,
+        h: &BlockHeader,
+    ) -> Result<(Vec<UnsignedMessage>, Vec<SignedMessage>), Error> {
+        chain::block_messages(self.cs.blockstore(), h).map_err(|err| err.into())
+    }
+
+    fn messages_for_tipset(&self, h: &Tipset) -> Result<Vec<UnsignedMessage>, Error> {
+        messages_for_tipset(self.cs.blockstore(), h).map_err(|err| err.into())
+    }
+
+    fn load_tipset(&self, tsk: &TipsetKeys) -> Result<Tipset, Error> {
+        self.cs.tipset_from_keys(tsk).map_err(|err| err.into())
+    }
+}
+
+/// This is the main MessagePool struct
+pub struct MessagePool<T: 'static> {
+    local_addrs: Vec<Address>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    pub cur_tipset: Arc<RwLock<Tipset>>,
+    api: Arc<RwLock<T>>,
+    pub min_gas_price: BigInt,
+    pub max_tx_pool_size: i64,
+    pub network_name: String,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    sig_val_cache: LruCache<Cid, ()>,
+    local_msgs: Vec<SignedMessage>,
+}
+
+impl<T> MessagePool<T>
+where
+    T: Provider + std::marker::Send + std::marker::Sync + 'static,
+{
+    /// Create a new message pool
+    pub async fn new(mut api: T, network_name: String) -> Result<MessagePool<T>, Error>
+    where
+        T: Provider,
+    {
+        // LruCache sizes have been taken from the lotus implementation
+        let pending = Arc::new(RwLock::new(HashMap::new()));
+        let tipset = Arc::new(RwLock::new(api.get_heaviest_tipset().ok_or_else(|| {
+            Error::Other("No ts in api to set as cur_tipset".to_owned())
+        })?));
+        let bls_sig_cache = Arc::new(RwLock::new(LruCache::new(40000)));
+        let sig_val_cache = LruCache::new(32000);
+        let api_mutex = Arc::new(RwLock::new(api));
+
+        let mut mp = MessagePool {
+            local_addrs: Vec::new(),
+            pending,
+            cur_tipset: tipset,
+            api: api_mutex,
+            min_gas_price: Default::default(),
+            max_tx_pool_size: 5000,
+            network_name,
+            bls_sig_cache,
+            sig_val_cache,
+            local_msgs: Vec::new(),
+        };
+
+        mp.load_local().await?;
+
+        let mut subscriber = mp.api.write().await.subscribe_head_changes();
+
+        let api = mp.api.clone();
+        let bls_sig_cache = mp.bls_sig_cache.clone();
+        let pending = mp.pending.clone();
+        let cur_tipset = mp.cur_tipset.clone();
+
+        task::spawn(async move {
+            loop {
```

Comment:
Just to note, this is not ideal to start up an infinite task every time `MessagePool::new` is called without a way to cancel it. Forces it to be used in a very specific way. Can revisit later I guess